src/paths/steiner.cpp

```diff
@@ -0,0 +1,586 @@
+            MATRIX(dp_cache, i, j) = MATRIX(distance, i, j);
+        }
+    }
+
+    /*
+     * Usage: Data structure to store index value of subsets.
+     * The index would be used in DP table.
+     */
+    dictionary subsetMap;
+    IGRAPH_CHECK(generateSubsets(&steiner_terminals_copy, igraph_vector_int_size(&steiner_terminals_copy), no_of_nodes, subsetMap, allSubsets));
+
+    int steiner_terminals_copy_size = igraph_vector_int_size(&steiner_terminals_copy);
+    for (igraph_integer_t m = 2; m <= steiner_terminals_copy_size; m++) {
+        for (igraph_integer_t i = 0; i < (igraph_integer_t)allSubsets.size(); i++) {
+            auto it = allSubsets.begin();
+            std::advance(it, i);
+            int_set D = *it;
+            igraph_integer_t indexOfSubsetD;
+            indexOfSubsetD = fetchIndexofMapofSets(D, subsetMap);
+
+            IGRAPH_ALLOW_INTERRUPTION();
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                MATRIX(dp_cache, indexOfSubsetD, j) = IGRAPH_INFINITY;
+            }
+
+            for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+                igraph_real_t distance1 = IGRAPH_INFINITY;
+
+                for (auto E :  D) {
+
+                    if (E != j) {
+                        igraph_real_t distanceEJ = MATRIX(distance, E, j);
+
+                        int_set DMinusE = D;
+
+                        /*
+                         *  A set with Singleton value E removed from subset D
+                         */
+                        for (auto item :  DMinusE) {
+                            if (item == E) {
+                                DMinusE.erase(item);
+                                break;
+                            }
+                        }
+
+                        igraph_integer_t indexOfSubsetDMinusE;
+                        if (DMinusE.size() == 1) {
+                            indexOfSubsetDMinusE = *DMinusE.begin();
+                        } else {
+                            indexOfSubsetDMinusE = fetchIndexofMapofSets(DMinusE, subsetMap);
+                        }
+
+                        if ((distanceEJ + MATRIX(dp_cache, indexOfSubsetDMinusE, j)) < distance1) {
+                            distance1 = distanceEJ + (MATRIX(dp_cache, indexOfSubsetDMinusE, j));
+                        }
+                    }
+                }
+
+                for (igraph_integer_t k = 0; k < no_of_nodes; k++) {
+                    MATRIX(dp_cache, indexOfSubsetD, k) = std::min(MATRIX(dp_cache, indexOfSubsetD, k), MATRIX(distance, k, j) + distance1);
+                }
+            }
+        }
+    }
+
+    igraph_real_t distance2 = IGRAPH_INFINITY;
+
+    for (igraph_integer_t j = 0; j < no_of_nodes; j++) {
+        igraph_real_t distance1 = IGRAPH_INFINITY;
+        IGRAPH_ALLOW_INTERRUPTION();
+        for (igraph_integer_t subset_C_iterator = 0; subset_C_iterator < no_of_terminals; subset_C_iterator++) {
```

Comment:
I'm a little surprised that `subset_C_iterator` is not actually an iterator :smile: